dojo/utils.py

```diff
@@ -50,55 +50,151 @@
+        engagement (:model:`dojo.Engagement`, optional): Engagement to filter findings by
+        test (:model:`dojo.Test`, optional): Test to filter findings by
+    """
+    if product:
+        custom_filter_type = 'product'
+        custom_filter = {'test__engagement__product': product}
+
+    elif engagement:
+        custom_filter_type = 'engagement'
+        custom_filter = {'test__engagement': engagement}
+
+    elif test:
+        custom_filter_type = 'test'
+        custom_filter = {'test': test}
+
+    else:
+        raise ValueError('No product, engagement or test provided as argument.')
+
+    deduplication_algorithm = finding.test.deduplication_algorithm
+
+    deduplicationLogger.debug(
+        'Matching finding %i:%s to existing findings in %s %s using %s as deduplication algorithm.',
+        finding.id, finding.title, custom_filter_type, list(custom_filter.values())[0], deduplication_algorithm
+    )
+
+    if deduplication_algorithm == 'hash_code':
+        return (
+            Finding.objects.filter(
+                **custom_filter,
+                hash_code=finding.hash_code
+            ).exclude(hash_code=None)
+            .exclude(id=finding.id)
+            .order_by('id')
+        )
+
+    elif deduplication_algorithm == 'unique_id_from_tool':
+        return (
+            Finding.objects.filter(
+                **custom_filter,
+                unique_id_from_tool=finding.unique_id_from_tool
+            ).exclude(unique_id_from_tool=None)
+            .exclude(id=finding.id)
+            .order_by('id')
+        )
+
+    elif deduplication_algorithm == 'unique_id_from_tool_or_hash_code':
+        query = Finding.objects.filter(
+            Q(**custom_filter),
+            (
+                (Q(hash_code__isnull=False) & Q(hash_code=finding.hash_code)) |
+                (Q(unique_id_from_tool__isnull=False) & Q(unique_id_from_tool=finding.unique_id_from_tool))
+            )
+        ).exclude(id=finding.id).order_by('id')
+        deduplicationLogger.debug(query.query)
+        return query
+
+    elif deduplication_algorithm == 'legacy':
+        # This is the legacy reimport behavior. Although it's pretty flawed and
+        # doesn't match the legacy algorithm for deduplication, this is left as is for simplicity.
+        # Re-writing the legacy deduplication here would be complicated and counter-productive.
+        # If you have use cases going through this section, you're advised to create a deduplication configuration for your parser
+        logger.debug("Legacy dedup. In case of issue, you're advised to create a deduplication configuration in order not to go through this section")
+        return (
+            Finding.objects.filter(
+                **custom_filter,
+                title=finding.title,
+                severity=finding.severity,
+                numerical_severity=Finding.get_numerical_severity(finding.severity)
```

Comment:
Default fields are specified here:
https://github.com/DefectDojo/django-DefectDojo/blob/386d2e69a4ada567e6b849e48b2978e22dcd19d1/dojo/models.py#L2442-L2445

Reply:
This one I didn't get.
You think we should change the filter by adding `cwe` + `line` + `file_path` + `description` and removing `severity` + `numerical_severity`?
The code is currently like this because it was based on the legacy reimport:
https://github.com/DefectDojo/django-DefectDojo/blob/dev/dojo/importers/reimporter/utils.py#L42C1-L52

Reply:
Hmm that is a good point. My thought was that if FP history was based on dedupe algorithms then the default  algorithm should based on the default dedupe algorithm.  

I suppose since true dedupe cannot be enabled with FP history, the chances of a "default algorithm mismatch" are pretty slim. I think you made the right call keeping the the same algorithm as the reimoprter

@mtesauro what are your thoughts?

Reply:
That is a very interesting point.

Given that you can have FP on and dedup off, relying on the legacy method seems to be the safest bet given how the code works today.

I agree with both of you (@Maffooch & @adiffpirate ) - lets go with the legacy method. :+1: